File: ios/SurfriderDataCards/Tally/Theme.swift
//
//  Nocturne, in Swift.
//
//  The design system's token sheet is `_ds/.../styles.css` in the design
//  project and the handoff lists every value the screens use. They are
//  transcribed here once and nothing else in the app writes a literal colour.
//
//  TWO SUBSTITUTIONS, both the ones the handoff asks for.
//
//  Type is Inter in the design and the system face here. The handoff's rule for
//  icons -- "on iOS, substitute the matching SF Symbol rather than bundling a
//  web font" -- is the same argument for the text: Inter and SF Pro are both
//  neutral grotesques at these sizes, and a bundled web font on iOS costs the
//  system's own optical sizing, Dynamic Type and the numeric variants. The
//  scale, the weights and the tracking below are the design's, exactly.
//
//  Icons are SF Symbols, named in `Icon` so the mapping from Phosphor is in one
//  place rather than spelled out at thirteen call sites.
//

import SwiftUI

enum Nocturne {

    // MARK: - Colour

    /// Page background.
    static let ground = Color(hex: 0x161826)
    /// Cards, inputs, keypad keys.
    static let surface = Color(hex: 0x232532)
    /// Never pure white.
    static let text = Color(hex: 0xE9E9ED)
    static let divider = text.opacity(0.16)

    /// Lines, marks, outlines -- never a flood.
    static let accent = Color(hex: 0x9184D9)
    /// Accent text at body size, where the accent itself would not carry.
    static let accent300 = Color(hex: 0xD2CEFD)
    /// Warnings.
    static let accent400 = Color(hex: 0xB5ABFC)
    /// Tinted borders.
    static let accent700 = Color(hex: 0x5D5294)
    /// Tag fills.
    static let accent800 = Color(hex: 0x423A6A)
    /// Tinted panel fills.
    static let accent900 = Color(hex: 0x2B2741)

    /// Progress track. The handoff names this "Neutral 800" and gives this hex;
    /// the token sheet has the same hex under `--color-neutral-900`. The hex is
    /// the thing being specified, so the hex is what is followed.
    static let track = Color(hex: 0x292B31)

    /// The capture screen darkens below the ground.
    static let captureGround = Color(hex: 0x0E0F18)

    /// Text at a fraction of full strength. The design expresses muted text as
    /// `color-mix(in srgb, text N%, transparent)`, which is this.
    static func text(_ percent: Double) -> Color { text.opacity(percent / 100) }

    // MARK: - Paper
    //
    // THE CROPS STAY ON WHITE ON PURPOSE. They are photographs of paper and the
    // job is reading faint pencil; inverting or dimming them costs contrast
    // exactly where it is scarcest, and misrepresents the card. They are framed
    // so they read as pictures rather than glare. Carried over verbatim from
    // the note at the top of src/style.css -- change it in both places or in
    // neither.

    enum Paper {
        static let fill = Color.white
        static let border = Color(hex: 0xC8CCD6)
        static let ruling = Color(hex: 0xE4E7EE)
        static let rule = Color(hex: 0xD5D9E2)
        static let ink = Color(hex: 0x26303C)
    }

    // MARK: - Spacing

    /// Horizontal page margin, everywhere.
    static let margin: CGFloat = 22
    /// The status bar overlays the content, so every screen starts below it.
    static let safeTop: CGFloat = 56
    /// Clears the home indicator.
    static let safeBottom: CGFloat = 46
    static let navBar: CGFloat = 44
    /// Nothing below this is a hit target.
    static let minTap: CGFloat = 44

    // MARK: - Radii

    enum Radius {
        static let viewfinder: CGFloat = 12
        static let draftCard: CGFloat = 10
        static let key: CGFloat = 9
        /// Inputs, buttons, panels, crops.
        static let base: CGFloat = 8
        static let thumb: CGFloat = 4
        static let pill: CGFloat = 999
    }

    // MARK: - Type
    //
    // Headings are weight 500 -- never heavier. Hierarchy is size and space.

    enum Face {
        /// Screen titles: 30 / 28 / 26.
        static func title(_ size: CGFloat) -> Font { .system(size: size, weight: .medium) }
        /// The item name on the checking screen.
        static let item = Font.system(size: 21, weight: .medium)
        /// The keypad entry display and the values in the card list.
        static func numeral(_ size: CGFloat) -> Font {
            .system(size: size, weight: .medium).monospacedDigit()
        }
        static func body(_ size: CGFloat = 13) -> Font { .system(size: size) }
        static func label(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
            .system(size: size, weight: weight)
        }
        static let cardTag = Font.system(size: 11, design: .monospaced)
    }

    // MARK: - Icons
    //
    // Phosphor regular in the design; the nearest SF Symbol here.

    enum Icon {
        static let add = "plus"                          // ph-plus
        static let back = "chevron.left"                 // ph-caret-left
        static let forward = "chevron.right"             // ph-caret-right
        static let capture = "camera"                    // ph-camera
        static let scan = "doc.text.viewfinder"          // capture, by PDF
        static let warning = "exclamationmark.triangle"  // ph-warning
        static let refused = "exclamationmark.circle"    // ph-warning-circle
        static let check = "checkmark"                   // ph-check
        static let pass = "checkmark.circle"             // ph-check-circle
        static let blocker = "xmark.circle"              // ph-x-circle
        static let privacy = "lock"                      // ph-lock-simple
        static let next = "arrow.right"                  // ph-arrow-right
        static let spreadsheet = "tablecells"            // ph-file-xls
        static let share = "square.and.arrow.up"         // ph-share-network
    }
}

// MARK: -

extension Color {
    /// `0x9184D9` reads like the design token it came from; `red:green:blue:`
    /// does not, and every one of these was transcribed from a hex string.
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255,
            opacity: 1
        )
    }
}

File: android/app/src/main/java/com/mateobesse/surfriderdatacards/tally/Theme.kt
//
//  Nocturne, in Kotlin.
//
//  The same tokens as ios/SurfriderDataCards/Tally/Theme.swift, transcribed
//  from the same handoff. Nothing else in the app writes a literal colour, so a
//  change to the design is a change here and in Theme.swift, or in neither.
//
//  THE SAME TWO SUBSTITUTIONS, for the same reason. Type is Inter in the design
//  and the system face here: the handoff's rule for icons -- use the platform's
//  own set rather than bundling a web font -- is the same argument for the
//  text, and the system face keeps font scaling and the numeric variants. The
//  scale, the weights and the tracking are the design's.
//
//  Icons are Material Symbols (outlined) in place of Phosphor, named in `Icon`
//  so the mapping is in one place, exactly as the SF Symbols are on iOS.
//
//  ONE PLATFORM DIFFERENCE, deliberately. There is no Liquid Glass here, so
//  every control is the flat Nocturne surface the design specifies -- the path
//  iOS takes below iOS 26. Glass.swift has no counterpart and needs none.
//

package com.mateobesse.surfriderdatacards.tally

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowForward
import androidx.compose.material.icons.automirrored.outlined.Backspace
import androidx.compose.material.icons.automirrored.outlined.KeyboardArrowLeft
import androidx.compose.material.icons.automirrored.outlined.KeyboardArrowRight
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Done
import androidx.compose.material.icons.outlined.DocumentScanner
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.material.icons.outlined.HighlightOff
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material.icons.outlined.PhotoCamera
import androidx.compose.material.icons.outlined.SaveAlt
import androidx.compose.material.icons.outlined.Share
import androidx.compose.material.icons.outlined.TableChart
import androidx.compose.material.icons.outlined.WarningAmber
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

object Nocturne {

    // Colour

    /** Page background. */
    val ground = Color(0xFF161826)
    /** Cards, inputs, keypad keys. */
    val surface = Color(0xFF232532)
    /** Never pure white. */
    val text = Color(0xFFE9E9ED)
    val divider = text.copy(alpha = 0.16f)

    /** Lines, marks, outlines -- never a flood. */
    val accent = Color(0xFF9184D9)
    /** Accent text at body size, where the accent itself would not carry. */
    val accent300 = Color(0xFFD2CEFD)
    /** Warnings. */
    val accent400 = Color(0xFFB5ABFC)
    /** Tinted borders. */
    val accent700 = Color(0xFF5D5294)
    /** Tag fills. */
    val accent800 = Color(0xFF423A6A)
    /** Tinted panel fills. */
    val accent900 = Color(0xFF2B2741)

    /** Progress track. See the note on `track` in Theme.swift. */
    val track = Color(0xFF292B31)

    /** The capture screen darkens below the ground. */
    val captureGround = Color(0xFF0E0F18)

    /**
     * Text at a fraction of full strength. The design expresses muted text as
     * `color-mix(in srgb, text N%, transparent)`, which is this.
     */
    fun text(percent: Int): Color = text.copy(alpha = percent / 100f)

    // Paper
    //
    // THE CROPS STAY ON WHITE ON PURPOSE. They are photographs of paper and the
    // job is reading faint pencil; inverting or dimming them costs contrast
    // exactly where it is scarcest. Carried over verbatim from src/style.css
    // and Theme.swift -- change it in all three places or in none.

    object Paper {
        val fill = Color.White
        val border = Color(0xFFC8CCD6)
        val ruling = Color(0xFFE4E7EE)
        val rule = Color(0xFFD5D9E2)
        val ink = Color(0xFF26303C)
    }

    // Spacing

    /** Horizontal page margin, everywhere. */
    val margin = 22.dp
    /**
     * Below the status bar. iOS carries 56pt of top padding because the status
     * bar overlays its content; here the status bar's own inset is applied
     * first and this is what the design has left over.
     */
    val belowStatusBar = 6.dp
    /** Above the navigation bar, on top of its inset. */
    val aboveNavBar = 16.dp
    /**
     * Nothing below this is a hit target. 48dp rather than the design's 44pt:
     * that is Android's accessibility minimum, and the design's number was
     * iOS's. The nav bar is the same height for the same reason -- its back
     * affordance is the whole of it.
     */
    val minTap = 48.dp
    val navBar = minTap

    object Radius {
        val viewfinder = 12.dp
        val draftCard = 10.dp
        val key = 9.dp
        /** Inputs, buttons, panels, crops. */
        val base = 8.dp
        val thumb = 4.dp
    }

    // Type
    //
    // Headings are weight 500 -- never heavier. Hierarchy is size and space.

    object Face {
        /** Screen titles: 30 / 28 / 26. Tracking is -1% of the size, as specified. */
        fun title(size: Int) = TextStyle(
            fontSize = size.sp,
            fontWeight = FontWeight.Medium,
            letterSpacing = (-size / 100f).sp,
            lineHeight = (size * 1.15f).sp,
        )

        /** The item name on the checking screen. */
        val item = TextStyle(
            fontSize = 21.sp,
            fontWeight = FontWeight.Medium,
            letterSpacing = (-0.21).sp,
            lineHeight = 26.sp,
        )

        /** The keypad entry display and the values in the card list. */
        fun numeral(size: Int) = TextStyle(
            fontSize = size.sp,
            fontWeight = FontWeight.Medium,
            fontFeatureSettings = "tnum",
        )

        /** Body copy: 13 at line-height 1.5, per the handoff. */
        fun body(size: Int = 13, lineHeight: Float = 1.5f) = TextStyle(
            fontSize = size.sp,
            lineHeight = (size * lineHeight).sp,
        )

        fun label(size: Int, weight: FontWeight = FontWeight.Normal, tracking: TextUnit = TextUnit.Unspecified) =
            TextStyle(fontSize = size.sp, fontWeight = weight, letterSpacing = tracking)

        val cardTag = TextStyle(fontSize = 11.sp, fontFamily = FontFamily.Monospace)
    }

    // Icons
    //
    // Phosphor regular in the design; the nearest outlined Material Symbol here.

    object Icon {
        val add = Icons.Outlined.Add                                  // ph-plus
        val back = Icons.AutoMirrored.Outlined.KeyboardArrowLeft      // ph-caret-left
        val forward = Icons.AutoMirrored.Outlined.KeyboardArrowRight  // ph-caret-right
        val capture = Icons.Outlined.PhotoCamera                      // ph-camera
        val scan = Icons.Outlined.DocumentScanner                     // capture, by PDF
        val warning = Icons.Outlined.WarningAmber                     // ph-warning
        val refused = Icons.Outlined.ErrorOutline                     // ph-warning-circle
        val check = Icons.Outlined.Done                               // ph-check
        val pass = Icons.Outlined.CheckCircle                         // ph-check-circle
        val blocker = Icons.Outlined.HighlightOff                     // ph-x-circle
        val privacy = Icons.Outlined.Lock                             // ph-lock-simple
        val next = Icons.AutoMirrored.Outlined.ArrowForward           // ph-arrow-right
        val spreadsheet = Icons.Outlined.TableChart                   // ph-file-xls
        val share = Icons.Outlined.Share                              // ph-share-network
        val save = Icons.Outlined.SaveAlt                             // keeping a copy
        val backspace = Icons.AutoMirrored.Outlined.Backspace         // delete.left
    }
}
